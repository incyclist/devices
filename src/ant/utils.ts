export const getBrand = (manId) => {
    if(manId===undefined)
        return "ANT+"

    switch (manId) {
        case 1: return "Garmin"; 
        case 16: return "Timex"; 
        case 23: return "Suunto";        
        case 52: return "Seiko";
        case 53: return "Seiko";
        case 70: return "Sigma";
        case 123: return "Polar"; 
        case 287: return "VDO"; 
        case 32: return "Wahoo"; 
        case 86: return "Elite"; 
        case 89: return "Tacx"; 
        default: 
            return "ANT+"
    }
}

