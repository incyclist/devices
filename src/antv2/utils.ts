import { Profile } from "incyclist-ant-plus"
import { LegacyProfile } from "./types"

const brands = [
    'Garmin,1',
    'garmin_fr405_antfs,2',
    'Zephyr,3',
    'Dayton,4',
    'idt,5',
    'SRM,6',
    'Quarq,7',
    'ibike,8',
    'Saris,9',
    'spark_hk,10',
    'Tanita,11',
    'echowell,12',
    'Dynastream_oem,13',
    'Nautilus,14',
    'Dynastream,15',
    'Timex,16',
    'Metrigear,17',
    'Xelic,18',
    'Beurer,19',
    'Cardiosport,20',
    'A&D,21',
    'Hmm,22',
    'Suunto,23',
    'Thita Elektronik,24',
    'Gpulse,25',
    'Clean Mobile,26',
    'Pedal Brain,27',
    'Peaksware,28',
    'Saxonar,29',
    'Lemond Fitness,30',
    'Dexcom,31',
    'Wahoo,32',
    'Octane,33',
    'Archinoetics,34',
    'The Hurt Box,35',
    'Citizen Systems,36',
    'Magellan,37',
    'Osynce,38',
    'Holux,39',
    'Concept2,40',
    'One Giant Leap,42',
    'Ace Sensor,43',
    'Brim Brothers,44',
    'Xplova,45',
    'Perception Digital,46',
    'Bf1systems,47',
    'Pioneer,48',
    'Spantec,49',
    'Metalogics,50',
    '4iiiis,51',
    'Seiko,52',
    'Seiko,53',
    'Ifor Powell,54',
    'Maxwell Guider,55',
    'Star rac,56',
    'Breakaway,57',
    'Alatech,58',
    'Mio,59',
    'Rotor,60',
    'Geonaute,61',
    'Id Bike,62',
    'Specialized,63',
    'Wtek,64',
    'Physical Enterprises,65',
    'North Pole,66',
    'Bkool,67',
    'Cateye,68',
    'Stages,69',
    'Sigma,70',
    'Tomtom,71',
    'Peripedal,72',
    'Wattbike,73',
    'Moxy,76',
    'Ciclosport,77',
    'Powerbahn,78',
    'Acorn,79',
    'Lifebeam,80',
    'Bontrager,81',
    'Wellgo,82',
    'Scosche,83',
    'Magura,84',
    'Woodway,85',
    'Elite,86',
    'Nielsen Kellerman,87',
    'DK City,88',
    'Tacx,89',
    'Direction,90',
    'Magtonic,91',
    '1partcarbon,92',
    'Inside Ride,93',
    'Sound Of Motion,94',
    'Stryd,95',
    'icg,96',
    'MiPulse,97',
    'Bsx Athletics,98',
    'Look,99',
    'Campagnolo,100',
    'Body Bike Smart,101',
    'Praxisworks,102',
    'Limits,103',
    'Topaction,104',
    'Cosinuss,105',
    'Fitcare,106',
    'Magene,107',
    'Giant,108',
    'Tigrasport,109',
    'Salutron,110',
    'Technogym,111',
    'Bryton,112',
    'Latitude,113',
    'Soaring,114',
    'Igpsport,115',
    'Thinkrider,116',
    'Gopher,117',
    'Waterrower,118',
    'Orangetheory,119',
    'Inpeak,120',
    'Kinetic,121',
    'Johnson,122',
    'Polar,123',
    'Seesense,124',
    'NCI,125',
    'Development,255',
    'Healthandlife,257',
    'Lezyne,258',
    'Scribe,259',
    'Zwift,260',
    'Watteam,261',
    'Recon,262',
    'Favero,263',
    'Dynovelo,264',
    'Strava,265',
    'Precor,266',
    'Bryton,267',
    'Sram,268',
    'Navman,269',
    'Cobi,270',
    'Spivi,271',
    'Mio Magellan,272',
    'Evesports,273',
    'Sensitivus Gauge,274',
    'Podoon,275',
    'Life Time Fitness,276',
    'Falco,277',
    'Minoura,278',
    'Cycliq,279',
    'Luxottica,280',
    'Trainer Road,281',
    'The Sufferfest,282',
    'Fullspeedahead,283',
    'Virtualtraining,284',
    'Feedbacksports,285',
    'Omata,286',
    'VDO,287',
    'Magneticdays,288',
    'Hammerhead,289',
    'Kinetic By Kurt,290',
    'Shapelog,291',
    'Dabuziduo,292',
    'Jetblack,293',
    'Coros,294',
    'Virtugo,295',
    'Velosense,296',
    'Cycligentinc,297',
    'Actigraphcorp,5759'
]

const brandLookup = brands.map( str => {
    const [name,id] = str.split(',')
    return {name,id:Number(id)}
})


export const getBrand = (manId:number) => {
    const found = brandLookup.find( l => l.id===manId)
    if (found)
        return found.name
    return;
}

export const mapLegacyProfile = ( legacy:LegacyProfile ):Profile => {
    switch (legacy){
        case 'Heartrate Monitor': return 'HR'
        case 'Power Meter': return 'PWR'
        case 'Smart Trainer': return 'FE'
        case 'Cadence Sensor':return 'CAD'
        case 'Speed Sensor': return 'SPD'
        case 'Speed + Cadence Sensor': return 'SC'
    }
} 

