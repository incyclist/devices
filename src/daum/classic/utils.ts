export function getCockpit( c) {
    switch ( c) {        
        case 10: 
            return "Cardio";
        case 20: 
            return "Fitness";
        case 30: 
            return "Vita De Luxe";
        case 40: 
            return "8008";
        case 50: 
            return "8080";
        case 60: 
            return "Therapie";
        default:
            return "Unknown";
    }
}

export function getBikeType( type) {
    if (type===undefined)
        return 0;

    switch (type) {
        case "triathlon":
            return 0;
        case "race":
            return 0;
        case "mountain":
            return 1;
        default:
            return 1;
    }
    
    // TODO: define possible values and parse (should be aligned with ANT* spec)
}


export function getAge(birthday) { // birthday is a date
    if (birthday===undefined) {
        return 30;
    }

    try {
        const bd = new Date(birthday);
        const ageDifMs = Date.now() - bd.getTime();
        var ageDate = new Date(ageDifMs); // miliseconds from epoch
        return Math.abs(ageDate.getUTCFullYear() - 1970);    
    } 
    catch (error) {
        return 30;
    }
}

export function getGender(sex) {
    if (sex===undefined)
        return 2;
    switch (sex) {
        case "M":
            return 1; 
        case "F":
            return 1; 
        default:
            return 2;
    }
    
}

export function getLength( length) {
    if (length===undefined || length===null)
        return 180; // average european

    let l = Math.round(length);

    if (l<100)
        return 100;
    if (l>220)
        return 220;

    return l;
}

export function getWeight(weight?) {
    if (weight===undefined || weight===null)
        return 80; // average european

    let m = Math.round(weight);
    if (isNaN(m))
        return 80;

    if (m<10)
        return 10;
    if (m>250)
        return 250;

    return m;
}

export function parseRunData( data) {
    const bikeData = {} as any
    bikeData.isPedalling = (data[4]>0);
    bikeData.power  = data[5]*5;
    bikeData.cadence = data[6];
    bikeData.speed = data[7];
    bikeData.heartrate = data[14];
    bikeData.distance = (data[9]*256+data[8])/10;
    bikeData.distanceInternal = (data[9]*256+data[8])*100;
    bikeData.time = (data[11]*256+data[10]);
    bikeData.gear = data[16];
    return bikeData
}


export function buildError ( status,err) {
    const message = err && err.message ? err.message: err;
    const error = new Error( message) as any;
    error.status = status;
    return error;
}

export function hexstr(arr,start?,len?) {
    var str = "";
    if (start===undefined) 
        start = 0;
    if ( len===undefined) {
        len = arr.length;
    }
    if (len-start>arr.length) {
        len = arr.length-start;
    }

    var j=start;
    for (var i = 0; i< len; i ++) {
        var hex = Math.abs( arr[j++]).toString(16);
        if ( i!==0 ) str+=" ";
        str+=hex;
    }
	return str;
}


export function Float32ToHex (float32)  {
    function getHex (i) { return ('00' + i.toString(16)).slice(-2) }
    var view = new DataView(new ArrayBuffer(4))
    view.setFloat32(0, float32);
    return Array.apply(null, { length: 4 }).map((_, i) => getHex(view.getUint8(i))).join('');
}

export function Float32ToIntArray (float32)  {
    var view = new DataView(new ArrayBuffer(4))
    view.setFloat32(0, float32);
    var arr = [];
  	for ( let i=0;i<4;i++) {
      arr.push( view.getUint8(i))
    }
    return arr;
}
