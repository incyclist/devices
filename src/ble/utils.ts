import { EventLogger } from "gd-eventlog";
import { LegacyProfile } from "../antv2/types";
import { BleCharacteristic,  BleProperty,  BleProtocol, BleRawPeripheral } from "./types";

type MappingRecord = {
    profile: LegacyProfile,
    protocol: BleProtocol,
}

export function mapLegacyProfile(profile:string):MappingRecord {
    switch (profile) {
        case 'Smart Trainer': return { profile:'Smart Trainer', protocol:'fm' }
        case 'Elite Smart Trainer': return { profile:'Smart Trainer', protocol:'elite' }
        case 'Elite SmartTrainer': return { profile:'Smart Trainer', protocol:'elite' }
        case 'Heartrate Monitor': return { profile:'Heartrate Monitor', protocol:'hr' }
        case 'Power Meter': return { profile:'Power Meter', protocol:'cp' }
        case 'Tacx Smart Trainer': return { profile:'Smart Trainer', protocol:'tacx' }
        case 'Tacx SmartTrainer': return { profile:'Smart Trainer', protocol:'tacx' }
        case 'Wahoo SmartTrainer': return { profile:'Smart Trainer', protocol:'wahoo' }
        case 'Wahoo Smart Trainer': return { profile:'Smart Trainer', protocol:'wahoo' }
    }
}

export function uuid (s:string):string {
    //console.log(s)
    if (s) {
        if (s.includes('-')) {
            const parts = s.split('-')
            const uuidNo = parseInt('0x'+parts[0])
            return uuidNo.toString(16).toLowerCase()
        }
        return s;
    }
}

export function matches (uuid1:string,uuid2:string):boolean {

    return parseUUID(uuid1)===parseUUID(uuid2)

}

export function getPeripheralInfo(p: BleRawPeripheral) {
    const {id,name,address,advertisement,services} = p;
    if (advertisement) {
        return {id,name:advertisement.localName,address,services:advertisement.serviceUuids}
    }
    else {
        return {id,name,address,services}
    }
}

export function getCharachteristicsInfo(c:BleCharacteristic) {
    const {uuid,properties,name,_serviceUuid} = c;

    const nameStr= name ? ` (${name})` : ''
    const serviceStr = _serviceUuid ? `${_serviceUuid}:` : ''

    return `${serviceStr}${uuid}${nameStr} ${properties}`
}




export const parseUUID = (str:string):string => {
    const uuid = str.toUpperCase()

    if (str.startsWith('0x')) { 
        const hex = uuid.slice(2)
        if (hex.length===4) { // 16bit)
            return `0000${hex}00001000800000805F9B34FB`    
        }
        if (hex.length===8) { // 32bit)
            return `${hex}00001000800000805F9B34FB`    
        }
    }
    else if (uuid.length===4) {
        return `0000${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===8) {
        return `${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===32) {
        return uuid
    }
    else if (uuid.length===36) {
        return uuid.replace(/-/g,'')

    }
    throw new Error(`Invalid UUID: ${uuid}`)
}

export const beautifyUUID = (str:string, withX:boolean = false ):string => {   

    let uuid

    try {
        uuid = parseUUID(str)
    }
    catch (err) {
        const logger = new EventLogger('Incyclist')
        logger.logEvent({message:'beautifyUUID error',uuid:str, error:err.message})
        return str
    }
    
    const parts = [
        uuid.substring(0,8),
        uuid.substring(8,12),    
        uuid.substring(12,16),
        uuid.substring(16,20),
        uuid.substring(20),
    ]

    if (uuid.substring(8)==='00001000800000805F9B34FB') {
        let short
        if (parts[0].startsWith('0000')) {
            short = parts[0].substring(4)
        }
        else {
            short  = parts[0]
        }
        return withX ? `0x${short}` : short
    }
    
    return parts.join('-')

}

export const fullUUID = (str:string):string => {   

    const uuid = parseUUID(str)
    
    const parts = [
        uuid.substring(0,8),
        uuid.substring(8,12),    
        uuid.substring(12,16),
        uuid.substring(16,20),
        uuid.substring(20),
    ]  
    return parts.join('-')

}


export const propertyVal = (properties:BleProperty[]):number => {
    let res = 0

    if (properties.includes('read'))    res |= 0x01
    if (properties.includes('write'))   res |= 0x02
    if (properties.includes('notify'))  res |= 0x04

    return res
}


export const propertyFromVal = (val:number):BleProperty[] => {
    const res:BleProperty[] = []

    if (val & 0x01) res.push('read')
    if (val & 0x02) res.push('write')   
    if (val & 0x04) res.push('notify')

    return res
}