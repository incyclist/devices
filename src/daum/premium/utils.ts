import { Route } from "../../types/route";
import { Gender, User } from "../../types/user";
import { DEFAULT_AGE, DEFAULT_USER_WEIGHT } from "../classic/utils";
import FileTime from 'win32filetime'

const sum = (arr) => arr.reduce( (a,b) => a+b,0);

export function bin2esc (arr) {
    if ( arr===undefined) {
        return;
    }

    const res = []
    arr.forEach( v => {
        switch (v) {
            case 0x12: res.push(0x22); res.push(0x12); break; 
            case 0x22: res.push(0x22); res.push(0x22); break; 
            case 0x01: res.push(0x22); res.push(0x11); break; 
            case 0x17: res.push(0x22); res.push(0x27); break; 
            case 0x06: res.push(0x22); res.push(0x16); break; 
            case 0x15: res.push(0x22); res.push(0x25); break; 
            default:
                res.push(v);

        }
    })
    return res
}

export function esc2bin(arr) {
    if ( arr===undefined) {
        return;
    }

    const res = []
    let escaped = false;
    arr.forEach( (v,idx) => {
        
        if (escaped) {
            escaped = false;
            switch (v) {
                case 0x11: res.push(0x1); return;
                case 0x27: res.push(0x17); return;
                case 0x16: res.push(0x6); return;
                case 0x25: res.push(0x15); return;
                case 0x12: res.push(0x12); return;
                case 0x22: res.push(0x22); return;
                default: res.push(v);
            }                
            return;
        }

        if ( v===0x22) {
            escaped = true;
        }
        else {
            res.push(v)
        }

        
    })
    return res

}


export function checkSum( cmdArr, payload) {

    const total = sum(cmdArr)+sum(payload);
    const checkSumVal = total % 100;
    let checkSumStr = checkSumVal.toString();
    while (checkSumStr.length<2) checkSumStr = '0'+checkSumStr;
    return checkSumStr;
}

export function buildMessage(command,payload?) {

    const cmdArr = getAsciiArrayFromStr(command);

    let data = payload || [];

    if ( typeof(payload) === 'number') 
        data = [payload]
    if ( typeof(payload) === 'string') 
        data = getAsciiArrayFromStr(payload)
    
    const checkSumVals = getAsciiArrayFromStr( checkSum( cmdArr, data) );

    const message = [];
    message.push(0x01);
    message.push(...cmdArr);
    if ( data.length>0)
        message.push(...data);
    message.push(...checkSumVals);
    message.push(0x17);
    return message;
}

export function getMessageData(command) {
    const res = [...command];
    res.splice(0,4); // remove SOH and header
    res.splice(-3,3);   // remove checksum and ETB
    return res;
}



export function hexstr(arr,start?,len?) {

    const isArr =  Array.isArray(arr) || arr instanceof Uint8Array;
    if(!arr) return '';

    var str = "";
    if (start===undefined) 
        start = 0;
    if ( len===undefined || (start+len>arr.length) ) {
        len = arr.length-start;
    }
    if(len<0) return '';

    var j=start;
    for (var i = 0; i< len; i ++) {
        const c = isArr ? arr[j++] : ascii(arr.charAt(j++))
        var hex = Math.abs(c).toString(16);
        if (hex.length<2) hex = '0'+hex;
        if ( i!==0 ) str+=" ";
        
        str+=hex;
    }
	return str;
}

export function getHex (i) { return ('00' + i.toString(16)).slice(-2) }

export function append (cmd, arr) { cmd.push(...arr)}



export function ascii(c) {
    if (c===undefined || c===null)
        return;

    return c.charCodeAt(0);
}

export function charArrayToString( arr) {
    if ( arr===undefined || arr==null) return undefined;

    let str = ''
    arr.forEach(c => str+=c);
    return str;
}

export function asciiArrayToString( arr) {
    if ( arr===undefined || arr==null) return undefined;

    let str = ''
    arr.forEach(c => str+= String.fromCharCode(c));
    return str;
}

export function getAsciiArrayFromStr (str) { 
    if (str===undefined || str===null)
        return undefined;
    
    const n = str.length;
    let result = [];
    for (let i=0;i<n;i++) {
        result.push( str.charCodeAt(i) );
    }
    return result;
}


export function Float32ToHex (float32)  {
    var view = new DataView(new ArrayBuffer(4))
    view.setFloat32(0, float32);
    return Array.apply(null, { length: 4 }).map((_, i) => getHex(view.getUint8(i))).join('');
}

export function Float32ToIntArray (float32)  {
    var view = new DataView(new ArrayBuffer(4))
    view.setFloat32(0, float32);
    var arr = [];
  	for ( let i=0;i<4;i++) {
      arr.push( view.getUint8(i))
    }
    return arr;
}


export function Int16ToIntArray (int16)  {
    var view = new DataView(new ArrayBuffer(2))
    view.setInt16(0, int16);
    var arr = [];
  	for ( let i=0;i<2;i++) {
      arr.push( view.getUint8(i))
    }
    return arr;
}

export function Int32ToIntArray (int32)  {
    var view = new DataView(new ArrayBuffer(4))
    view.setInt32(0, int32);
    var arr = [];
  	for ( let i=0;i<4;i++) {
      arr.push( view.getUint8(i))
    }
    return arr;
}


export enum ReservedCommands {
    RESULT_RESET = 0,
    RESULT_GET = 1,
    NETRACE_START = 2,
    NETRACE_STOP = 3,
    NETRACE_USERNAME = 4,
    NETRACE_USERDATA = 5,
    PERSON_GET = 6,
    PERSON_SET = 7,
    PROGRAM_LIST_BEGIN = 8,
    PROGRAM_LIST_NEW_PROGRAM = 9,
    PROGRAM_LIST_CONTINUE_PROGRAM = 10,
    PROGRAM_LIST_END = 11,
    PROGRAM_LIST_START = 12,
    RELAX_START = 12,
    RELAX_STOP = 14,
    RELAX_GET_DATA = 0xF,
    KEY_PRESSED = 0x10,
    PROGRAM_CONTROL = 0x11
}

export enum BikeType {
    ALLROUND = 0,
    RACE = 1,
    MOUNTAIN =2
}

export function getBikeType( bikeTypeStr?:string): BikeType {
    if (bikeTypeStr===undefined || bikeTypeStr===null)
        return BikeType.RACE;

    if (bikeTypeStr.toLowerCase()==='mountain') 
        return BikeType.MOUNTAIN;
    
    return BikeType.RACE;
}

export function routeToEpp(route:Route, date?:Date): Uint8Array {
    
    const buffer = Buffer.alloc( 376 + route.points.length*12 );

    const fileTime = FileTime.fromUnix( date? date: Date.now())
    
    let offset = 0;

    const name = route.name || ''
    const description = route.description || ''
    const minElevation = route.minElevation ? route.minElevation : 0;
    const maxElevation = route.maxElevation ? route.minElevation : Math.max ( ...route.points.map(p => p.elevation))
    const sampleRate = route.points.length!==0 ? Math.round(route.totalDistance/route.points.length) : 0;

    buffer.writeUInt32LE(fileTime.low,offset);offset+=4;
    buffer.writeUInt32LE(fileTime.high,offset); offset+=4;
    buffer.write(name ,offset, name.length, 'ascii'); offset +=64;
    buffer.write(description,offset, description.length, 'ascii'); offset +=256;

    buffer.writeInt32LE(0x10,offset);offset+=4;                                     // ProgramType = DISTANCE_HEIGHT=0x10
    buffer.writeInt32LE(0x0,offset);offset+=4;                                      // reserved
    buffer.writeInt32LE(minElevation,offset);offset+=4;                             // minElevation
    buffer.writeInt32LE(maxElevation,offset);offset+=4;                             // maxElevation
    buffer.writeInt32LE(route.points.length,offset);offset+=4;                      // points count
    buffer.writeInt32LE(sampleRate,offset);offset+=4;                               // sample rate
    buffer.writeInt32LE(0x01,offset);offset+=4;                                     // valid for ( BITs: 1: bike, 2: lyps, 4: run)
    buffer.writeInt32LE(0x0,offset);offset+=4;                                      // elevation start      // TODO: why 0 and not 2st elevation ??
    buffer.writeInt16LE(0x0,offset);offset+=2;                                      // power limit
    buffer.writeInt16LE(0x0,offset);offset+=2;                                      // heart rate limit
    buffer.writeInt32LE(0x0,offset);offset+=4;                                      // speed limit
    buffer.writeInt32LE(0x0,offset);offset+=4;                                      // reserved
    buffer.writeInt32LE(0x0,offset);offset+=4;                                      // reserved
    
    route.points.forEach(p => {
        buffer.writeUInt32LE(sampleRate,offset);offset+=4;                          // sample rate
        buffer.writeFloatLE(p.elevation,offset);offset+=4;                          // elevation          
        buffer.writeFloatLE(0,offset);offset+=4;                                    // reserved
    })

    return new Uint8Array(buffer);
}



export function parseTrainingData(payload) {
    
    const GS = 0x1D

    const speedVals=['ok','too low','too high']
    const gearVal = (v) => v>0 ? v-1 : undefined;

    //const strVals = payload.reduce ( (str,c) => str + (c==0x1d ? '#' :String.fromCharCode(c) ),'')
    const vals = payload.split( String.fromCharCode(GS) );
    
    const data = {
        time: parseInt(vals[0]),
        heartrate: parseInt(vals[1]),
        speed: parseFloat(vals[2]) *3.6,
        slope: parseFloat(vals[3]),
        distanceInternal: parseInt(vals[4]),
        cadence: parseFloat(vals[5]),
        power: parseInt(vals[6]),
        physEnergy: parseFloat(vals[7]),
        realEnergy: parseFloat(vals[8]),
        torque: parseFloat(vals[9]),
        gear:  gearVal(parseInt(vals[10])),
        deviceState: parseInt(vals[11]),
        speedStatus: speedVals[parseInt(vals[12])],
    }

    return data;
 
}

	// Data Structure:
    // ---------------------------------------------------------
	// PMWTL_PERSONAL_DATA					136 == 0x88
	// TMWTL_PERSONAL_DATA
	// 
	// LIMIT_COMPARE_STRUCT limits			112

    	/// LIMIT_COMPARE_STRUCT				112
        /// 
        /// LMM_STRUCT limit					56
        /// LMM_STRUCT limitCompare				56

            /// LMM_STRUCT							56
            /// 
            /// LIMIT_STRUCT min					28
            /// LIMIT_STRUCT max					28


                /// LIMIT_STRUCT						28
                /// 
                /// TFLOAT speed						4
                /// TFLOAT slope;						4
                /// TFLOAT distance;					4
                /// TFLOAT nm;							4
                /// TUINT16 height;						2
                /// TUINT16 kJoule;						2
                /// TUINT16 watt;						2
                /// TUINT16 forbiddenTime;				2
                /// TUINT8 pulse;						1
                /// TUINT8 time;						1
                /// TUINT8 rpm;							1
                /// TUINT8 unused1;						1
                /// 
                /// Hint: All can be disabled with value 0, except for slope, which needs to be disabled with -100        

	// 
	// TSINT32 sex							4	// 1 = male, 2 = female
	// TSINT32 age							4   // 00
	// TSINT32 height						4   // 181
	// TFLOAT weight						4   // float: 00 00 b1 42
	// TFLOAT realisticKJFactor			    4   // 0
	// TUINT32 cockpitPerson				4   // 0

export function getPersonData(user: User) {

    const buffer =Buffer.alloc(136);
    let offset = 0;

    // write limit compare struct
    for (let i=0;i<4;i++) {
        buffer.writeInt32LE(0,offset); offset+=4;           // speed = 0;
        buffer.writeFloatLE(-100,offset); offset+=4;        // slope = -100;
        buffer.writeFloatLE(0,offset); offset+=4;           // distance = 0;
        buffer.writeFloatLE(0,offset); offset+=4;           // nm = 0;
        buffer.writeUInt16LE(0,offset); offset+=2;          // height = 0;
        buffer.writeUInt16LE(0,offset); offset+=2;          // kJoule = 0;
        buffer.writeUInt16LE(0,offset); offset+=2;          // watt = 0;
        buffer.writeUInt16LE(0,offset); offset+=2;          // forbiddenTime = 0;
        buffer.writeUInt8(0,offset); offset+=1;             // pulse = 0;
        buffer.writeUInt8(0,offset); offset+=1;             // time = 0;
        buffer.writeUInt8(0,offset); offset+=1;             // rpm = 0;
        buffer.writeUInt8(0,offset); offset+=1;             // unused1 = 0;
    }

    buffer.writeInt32LE( 1   /*user.sex=== Gender.FEMALE ? 2: 1*/, offset );offset+=4;
    buffer.writeInt32LE( 1   /*user.age!==undefined ? user.age :  1*/, offset);offset+=4;
    buffer.writeInt32LE( 175 /*user.length!==undefined ? user.length :  175*/, offset);offset+=4;
    buffer.writeFloatLE( 70  /*user.weight!==undefined ? user.weight :  70*/, offset);offset+=4;
    buffer.writeFloatLE( 0,offset);offset+=4;               // realisticKJFactor = 0
    buffer.writeUInt32LE( 1,offset);offset+=4;              // cockpitPerson = 1;
    
    return buffer;

}
