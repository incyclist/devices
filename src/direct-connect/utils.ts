import { BleProperty } from "../ble/types"

export const parseUUID = (str:string):string => {
    const uuid = str.toUpperCase()

    if (str.startsWith('0x')) { 
        const hex = uuid.slice(2)
        if (hex.length===4) { // 16bit)
            return `0000${hex}00001000800000805F9B34FB`    
        }
        if (hex.length===8) { // 32bit)
            return `${hex}00001000800000805F9B34FB`    
        }
    }
    else if (uuid.length===4) {
        return `0000${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===8) {
        return `${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===32) {
        return uuid
    }
    else if (uuid.length===36) {
        return uuid.replace(/-/g,'')

    }
    throw new Error(`Invalid UUID: ${uuid}`)
}

export const beautifyUUID = (str:string, withX:boolean = false ):string => {   

    const uuid = parseUUID(str)
    
    const parts = [
        uuid.substring(0,8),
        uuid.substring(8,12),    
        uuid.substring(12,16),
        uuid.substring(16,20),
        uuid.substring(20),
    ]

    if (uuid.substring(8)==='00001000800000805F9B34FB') {
        let short
        if (parts[0].startsWith('0000')) {
            short = parts[0].substring(4)
        }
        else {
            short  = parts[0]
        }
        return withX ? `0x${short}` : short
    }
    
    return parts.join('-')

}

export const propertyVal = (properties:BleProperty[]):number => {
    let res = 0

    if (properties.includes('read'))    res |= 0x01
    if (properties.includes('write'))   res |= 0x02
    if (properties.includes('notify'))  res |= 0x04

    return res
}


export const propertyFromVal = (val:number):BleProperty[] => {
    const res:BleProperty[] = []

    if (val & 0x01) res.push('read')
    if (val & 0x02) res.push('write')   
    if (val & 0x04) res.push('notify')

    return res
}
