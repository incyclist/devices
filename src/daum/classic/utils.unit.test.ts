import { getWeight, parseRunData } from "./utils"

describe ('utils',()=>{

    test( '50kg',()=>{
        const res=getWeight(50)
        expect(res).toBe(50)
    })

    test( '50.123kg => returns 50',()=>{
        const res=getWeight(50.123)
        expect(res).toBe(50)
    })

    test( '49.5kg => returns 50',()=>{
        const res=getWeight(49.5)
        expect(res).toBe(50)
    })

    test( '<10kg => returns 10',()=>{
        const res=getWeight(2)
        expect(res).toBe(10)
    })

    test( '<250kg => returns 250',()=>{
        const res=getWeight(2000)
        expect(res).toBe(250)
    })

    test( 'undefined => returns 80',()=>{
        const res=getWeight(undefined)
        expect(res).toBe(80)
    })
    test( 'null => returns 80',()=>{
        const res=getWeight(null)
        expect(res).toBe(80)
    })

    test( 'string => returns 80',()=>{
        const res=getWeight('John Doe')
        expect(res).toBe(80)
    })

    test( 'number as string => returns number (rounded)',()=>{
        const res=getWeight('75.2')
        expect(res).toBe(75)
    })

    describe ('parseRunData',()=> { 
        test( 'valid data' ,()=>{ 
            let error=undefined;
            let data = undefined;
            try {
                data = parseRunData( [0x40,0,0,0,0xC0,0x13,0,0,0xF9,3,0x28,0x2E,3,41,0,0,7,0xE4,0x32]);
            } catch (e) {
                error =e;
            }
            expect(error).toBeUndefined();
            expect(data).toBeDefined()
            expect(data).toMatchObject( {power:95})
        })

        test( 'invalidvalid data: "pedalling=0x40"' ,()=>{ 
            let error=undefined;
            let data = undefined;
            try {
                data = parseRunData( [0x40,0,0,0,0x40,5,0,0,0,0,0,0,0,0,0,0,0xA,0,0]);
            } catch (e) {
                error =e;
            }
            expect(error).toBeUndefined();
            expect(data).toBeDefined()
            expect(data).toMatchObject( {power:25})

            //expect(error).toBeDefined();
            //expect(error.message).toBe('Invalid data')
        })
    })

})