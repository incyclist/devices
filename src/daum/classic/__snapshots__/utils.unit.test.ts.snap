// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils Float32ToHex Integers 1`] = `
Array [
  "0:00000000",
  "1:3F800000",
  "2:40000000",
  "3:40400000",
  "4:40800000",
  "5:40A00000",
  "6:40C00000",
  "7:40E00000",
  "8:41000000",
  "9:41100000",
  "10:41200000",
  "11:41300000",
  "12:41400000",
  "13:41500000",
  "14:41600000",
  "15:41700000",
  "16:41800000",
  "17:41880000",
  "18:41900000",
  "19:41980000",
  "20:41A00000",
  "21:41A80000",
  "22:41B00000",
  "23:41B80000",
  "24:41C00000",
  "25:41C80000",
  "26:41D00000",
  "27:41D80000",
  "28:41E00000",
  "29:41E80000",
  "30:41F00000",
  "31:41F80000",
  "32:42000000",
  "33:42040000",
  "34:42080000",
  "35:420C0000",
  "36:42100000",
  "37:42140000",
  "38:42180000",
  "39:421C0000",
  "40:42200000",
  "41:42240000",
  "42:42280000",
  "43:422C0000",
  "44:42300000",
  "45:42340000",
  "46:42380000",
  "47:423C0000",
  "48:42400000",
  "49:42440000",
  "50:42480000",
  "51:424C0000",
  "52:42500000",
  "53:42540000",
  "54:42580000",
  "55:425C0000",
  "56:42600000",
  "57:42640000",
  "58:42680000",
  "59:426C0000",
  "60:42700000",
  "61:42740000",
  "62:42780000",
  "63:427C0000",
  "64:42800000",
  "65:42820000",
  "66:42840000",
  "67:42860000",
  "68:42880000",
  "69:428A0000",
  "70:428C0000",
  "71:428E0000",
  "72:42900000",
  "73:42920000",
  "74:42940000",
  "75:42960000",
  "76:42980000",
  "77:429A0000",
  "78:429C0000",
  "79:429E0000",
  "80:42A00000",
  "81:42A20000",
  "82:42A40000",
  "83:42A60000",
  "84:42A80000",
  "85:42AA0000",
  "86:42AC0000",
  "87:42AE0000",
  "88:42B00000",
  "89:42B20000",
  "90:42B40000",
  "91:42B60000",
  "92:42B80000",
  "93:42BA0000",
  "94:42BC0000",
  "95:42BE0000",
  "96:42C00000",
  "97:42C20000",
  "98:42C40000",
  "99:42C60000",
]
`;
